docs/proposals/scheduling/multi-podtemplate-support/multiple-pod-template-support.md

```diff
@@ -0,0 +1,195 @@
+// one component, associated with the podTemplate in the Deployment definition.
+//
+// Complex resources can have multiple components controlled through different podTemplates.
+// Each replica for the resource will fall into a component type with requirements defined by its relevant podTemplate.
+type ComponentRequirements struct {
+
+    // Name of this component
+    Name string `json:"name,omitempty"`
+
+    // Replicas represents the replica number of the resource's component
+    // +optional
+	Replicas int32 `json:"replicas,omitempty"`
+
+    // ReplicaRequirements represents the requirements required by each replica for this component.
+	// +optional
+	ReplicaRequirements *ReplicaRequirements `json:"replicaRequirements,omitempty"`
+
+}
+
+// ReplicaRequirements represents the requirements required by each replica.
+type ReplicaRequirements struct {
+
+	// NodeClaim represents the node claim HardNodeAffinity, NodeSelector and Tolerations required by each replica.
+	// +optional
+	NodeClaim *NodeClaim `json:"nodeClaim,omitempty"`
+
+	// ResourceRequest represents the resources required by each replica.
+	// +optional
+	ResourceRequest corev1.ResourceList `json:"resourceRequest,omitempty"`
+
+	// Namespace represents the resources namespaces
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// PriorityClassName represents the components priorityClassName
+	// +optional
+	PriorityClassName string `json:"priorityClassName,omitempty"`
+
+}
+```
+
+### Overview of Code Changes
+
+Besides the change to the `ResourceBindingSpec` and the `ReplicaRequirements API`, we will need to make a code change to the accurate estimator's implementation,
+which can be found here: https://github.com/karmada-io/karmada/blob/5e354971c78952e4f992cc5e21ad3eddd8d6716e/pkg/estimator/server/estimate.go#L59, as well as the maxReplica estimation done by the ResourceQuota plugin.
+
+Currently the accurate estimator will calculate the maxReplica count by:
+1. Running the maxReplica calculation (`maxReplicas`) for each plugin enabled by the accurate estimator.
+2. The accurate estimator will then loop through all nodes and sum up the amount of replicas (`sumReplicas`) that can fit in each node. This is to account for the resource fragmentation issue.
+3. The result returned will be: `Math.min(maxReplicas, sumReplicas)`.
+
+![Accurate-Scheduler-Steps](Accurate-Scheduler-Steps.png)
+
+For the proposed implementation, please refer to the next section.
+
+### Accurate Estimator Changes
+
+`Assumption 1`: Resources with more than one replica will always be scheduled to the same cluster.
+	- This simplifies the scope of the problem, and accounts for the fact that it is  non-trivial to schedule components of the same CRD across multiple clusters.
+
+`Assumption 2`: MaxReplica estimation will use a sum of all the resource requirement for every component's replica.
```

Comment:
So please correct me if I was wrong，you mean that we consider all components(podTemplate) together to compute how many CRDs a cluster can deploy?The assign result is a replica number of a whole crd，not one pod，right？What I am concerned about is what the targetCluster field of `RB` returns, whether it remains the same or returns different replicas for different podtemplates.
And if that was true, is it not considered that multiple pods under one crd are deployed on different nodes?

Reply:
Will try to answer these questions, let me know if anything else needs clarification:

> we consider all components(podTemplate) together to compute how many CRDs a cluster can deploy

Correct, but just for the estimation portion. If we sum all requirements together we can estimate how many multiples of the entire CRD can be scheduled on the cluster's available resources (even if that ends up being an overestimate due to resource-partitioning). 

> assign result is a replica number of a whole crd，not one pod，right

The assign result for the resource will still be in terms of pods, since that's the atomic scheduling unit in Karmada. In terms of the targetCluster field I believe it would be the same for both podTemplates, since this design assumes that CRDs with multiple pod templates want to be scheduled to the same cluster. 

Providing an example for FlinkDeployment (for simplicity lets just estimate CPU capacity): 

```
JobManagerPodTemplate
   replicas: 1
   replicaRequirements:
       cpu: 1
TaskManagedPodTemplate
   replicas: 2
   replicaRequirements:
       cpu: 1
```

Let's assume Cluster A has 7 available CPU. For estimation purposes we sum up the total CPU = (1 + 2) = 3. Total # of CRDs that can fit will be 7 / 3 = 2. 

Once we verify that all individual components can be scheduled on available nodes (to account for resource fragmentation), we return the estimation in terms of replicas = 2 * (3) = 6 replicas. 

The Karmada scheduler will then assign the 3 total replicas of the CRD to the 6 available replicas in Cluster A. 

> And if that was true, is it not considered that multiple pods under one crd are deployed on different nodes?

Well it is possible that different pods of the same CRD are deployed on different nodes. Unless you require something like gang-scheduling, where all pods must be scheduled on the same node. Is that something you require?